### PR TITLE
Milestone5: Database Performance Audit.

### PR DESCRIPTION
**Pull request by carlylechia**
- For this milestone, my partner and I started by adding so much data into the visits and animals tables, hence making them have several million rows each, according to the [project requirements](https://github.com/microverseinc/curriculum-databases/blob/main/db-structure/vet_clinic_performance_audit.md).

- Being a performance audit milestone, we then wrote a few (three) queries which each took more than 1000ms to execute. Together, we then brainstormed ways to make the queries get executed faster, hence making our database more performant. Screenshots of the query analyses were taken before and after the manipulations as evidence of the performance enhancements. This code was recorded in the `queries.sql` file, while the other code which made changes to the schema and updated the data were recorded in the `schema.sql` and `data.sql` files respectively.
- The screenshots are attached below:

- [x] **`EXPLAIN ANALYZE SELECT COUNT(*) FROM visits WHERE animal_id = 4;`**
- BEFORE:
1a

- AFTER:
1b

- [x] **`EXPLAIN ANALYZE SELECT * FROM visits WHERE vet_id = 2;`**
- BEFORE:
2a

- AFTER:
2b

- [x] **`EXPLAIN ANALYZE SELECT * FROM owners WHERE email = 'owner_18327@mail.com';`**
- BEFORE:
3a

- AFTER:
3b

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -163,3 +163,62 @@ SELECT species.name AS expected_specialty FROM animals JOIN visits ON animals.id
...
+--    ->  Gather  (cost=170348.08..170348.29 rows=2 width=8) (actual time=1000.251..1003.862 rows=3 loops=1)
+--          Workers Planned: 2
+--          Workers Launched: 2
+--          ->  Partial Aggregate  (cost=169348.08..169348.09 rows=1 width=8) (actual time=981.041..981.043 rows=1 loops=3)
+--                ->  Parallel Seq Scan on visits  (cost=0.00..167854.46 rows=597446 width=0) (actual time=0.082..917.412 rows=479238 loops=3)
+--                      Filter: (animal_id = 4)
+--                      Rows Removed by Filter: 4792380
+--  Planning Time: 0.070 ms
+--  Execution Time: 1003.925 ms
+CREATE INDEX idx_animal_id ON visits(animal_id);
```
Comment (Kingstalux) [inline queries.sql]: - [ ] Great job so far 👍 The execution time of your queries has been greatly improved 💯 Kindly put your solutions for improving the slow queries in the `schema.sql` file as per the requirements 👍 
- Requirements:
Annotation 2022-05-17 081857

File: queries.sql [hunk 2]
```diff
@@ -163,3 +163,62 @@ SELECT species.name AS expected_specialty FROM animals JOIN visits ON animals.id
...
+--                      Heap Fetches: 0
+--  Planning Time: 0.389 ms
+--  Execution Time: 119.881 ms
+EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;
+-- Seq Scan on visits  (cost=0.00..283171.65 rows=3877274 width=12) (actual time=1.745..1251.164 rows=3953717 loops=1)
+--    Filter: (vet_id = 2)
+--    Rows Removed by Filter: 11861135
+--  Planning Time: 0.101 ms
+--  Execution Time: 1463.933 ms
+CREATE INDEX idx_vet_id_asc ON visits (vet_id ASC);
```
Comment (Kingstalux) [inline queries.sql]: - [ ] Same issue here as above 👍

File: queries.sql [hunk 3]
```diff
@@ -163,3 +163,62 @@ SELECT species.name AS expected_specialty FROM animals JOIN visits ON animals.id
...
+EXPLAIN ANALYZE SELECT * FROM owners where email = 'owner_18327@mail.com';
+--  Gather  (cost=1000.00..142496.07 rows=4 width=43) (actual time=9.378..3144.924 rows=4 loops=1)
+--    Workers Planned: 2
+--    Workers Launched: 2
+--    ->  Parallel Seq Scan on owners  (cost=0.00..141495.67 rows=2 width=43) (actual time=1831.858..3133.374 rows=1 loops=3)
+--          Filter: ((email)::text = 'owner_18327@mail.com'::text)
+--          Rows Removed by Filter: 3333334
+--  Planning Time: 2.373 ms
+--  Execution Time: 3144.978 ms
+CREATE INDEX idx_email ON owners (email);
```
Comment (Kingstalux) [inline queries.sql]: - [ ] Same issue here as above 👍

**Review by Kingstalux — CHANGES_REQUESTED** [review]
Hi @carlylechia and @samiullahbahadur, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
149637855-7722d04c-b45d-4185-925f-7de047106461


## Highlights 👏 
✅ Queries have been optimized 🔥
✅ Great pull request description 👍 
✅ Professional README file 💯 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by julie-ify — APPROVED** [review]
Hi @carlylechia and @samiullahbahadur, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Status: Approved ✔️ 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**